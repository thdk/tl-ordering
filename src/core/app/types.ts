import { Action } from "redux";
import { IOrderItem, IOrderItemState } from "../orderitems/types";
import { IOrder, IOrderState } from "../orders/types";
import { IProductState } from "../products/types";

export interface IState {
    orders: IOrderState;
    orderItems: IOrderItemState;
    products: IProductState;
}

export interface IAction<T, P = undefined> extends Action<T> {
    payload?: P;
}

export interface IOrderWithData extends IOrder {
    items: IOrderItemWithPrice[];
    total: number;
}

export interface IOrderItemWithPrice extends IOrderItem {
    readonly unitPrice: number;
}
