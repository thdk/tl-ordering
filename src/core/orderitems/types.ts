export interface IOrderItemDictionary { [orderId: string]: IOrderItem[]; }

export interface IOrderItemState { byOrderId: IOrderItemDictionary; }

export interface IOrderItem {
    readonly productId: string;
    readonly quantity: number;
}

// Types for deserialized api data

export interface IOrderItemData {
    productId: string;
    quantity: number;
    orderId: string;
}

// Types representing data from/for API

interface IApiData {
    [manyProps: string]: any;
}

// "product-id": "B102",
// "quantity": "5",
// "unit-price": "4.99",
// "total": "24.95"
export interface IApiOrderItem extends IApiData {
    quantity: string;
    total: string;
}
