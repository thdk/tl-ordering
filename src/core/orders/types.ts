import { IApiOrderItem, IOrderItemData } from "../orderitems/types";

export interface IOrderState {
    byId: IOrderDictionary;
    visibleIds: string[];
    isLoading: boolean;
}

export interface IOrderDictionary { [orderId: string]: IOrder; }

export interface IOrder {
    readonly id: string;
    readonly customerId: string;
    readonly isPlaced?: boolean;
}

// Types for deserialized api data

export interface IOrderData {
    id: string;
    items: IOrderItemData[];
    customerId: string;
}

export interface IPlaceOrderData {
    success: boolean;
    order: IOrderData;
    reason?: string;
}

// Types representing data from/for API

interface IApiData {
    [manyProps: string]: any;
}

// {
//  "id": "2",
//   "customer-id": "2",
//   "items": [
//     {
//       "product-id": "B102",
//       "quantity": "5",
//       "unit-price": "4.99",
//       "total": "24.95"
//     }
//   ],
//   "total": "24.95"
// }
export interface IApiOrder extends IApiData {
    id: string;
    items: IApiOrderItem[];
    total: string;
}

export interface IApiPlaceOrderResult extends IApiData {
    result: string;
    order: IApiOrder;
    reason?: string;
}
