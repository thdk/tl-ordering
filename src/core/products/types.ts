export interface IProductDictionary { [productId: string]: IProduct; }

export interface IProductState {
    byId: IProductDictionary;
    allIds: string[];
    isLoading: boolean;
}

export enum ProductCategory {
    "unknown" = 0,
    "tools" = 1,
    "electronics" = 2,
}

export interface IProduct {
    readonly id: string;
    readonly description: string;
    readonly category: ProductCategory;
    readonly price: number;
}

// Types for deserialized api data

export interface IProductData {
    readonly id: string;
    readonly description: string;
    readonly category: ProductCategory;
    readonly price: number;
}

// Types representing data from/for API

export interface IApiProduct {
    readonly id: string;
    readonly description: string;
    readonly category: string;
    readonly price: string;
}
