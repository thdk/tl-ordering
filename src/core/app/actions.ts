import { IOrderAction } from "../core/orders/actions";
import { IOrderItemAction } from "./orderItems";

export type AnyAction = IOrderAction | IOrderItemAction;
