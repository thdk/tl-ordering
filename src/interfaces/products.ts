export interface IProduct {
    id: string;
    unitPrice: number;
}