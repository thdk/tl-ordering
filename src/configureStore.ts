import { applyMiddleware, createStore } from "redux";
import logger from "redux-logger";
import thunkMiddleware from "redux-thunk";

import rootReducer from "./core/app/reducer";
import { IState } from "./core/app/types";

export function configureStore(initialState: Partial<IState>) {
    return createStore(
        rootReducer,
        initialState,
        applyMiddleware(
            thunkMiddleware,
            logger, // TODO: only add logger middleware in development
        ),
    );
}
