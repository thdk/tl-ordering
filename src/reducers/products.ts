import { IProductState } from "../interfaces/state";

export function products(state: IProductState = { byId: {}, allIds: [] }) {
    return state;
}