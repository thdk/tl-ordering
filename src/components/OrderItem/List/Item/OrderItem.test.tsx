import React from "react";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import renderer from "react-test-renderer";
import configureStore, { MockStoreEnhanced } from "redux-mock-store";
import { IState } from "../../../../core/app/types";
import { ProductCategory } from "../../../../core/products/types";

import { shallow, ShallowWrapper } from "enzyme";
import ConnectedOrderItem, { OrderItemListItem } from "./OrderItem";

const mockStore = configureStore<Partial<IState>>();

const initialState: Pick<IState, "products"> = {
   products: {
      byId: {
         p1: {
            price: 10.99,
            description: "some product",
            category: ProductCategory.tools,
            id: "p1",
         },
      },
      allIds: ["p1"],
      isLoading: false,
   },
};

describe("OrderItem rendering for a given state from redux store:", () => {
   let store: MockStoreEnhanced;
   beforeEach(() => {
      store = mockStore(initialState);
      store.dispatch = jest.fn();
   });

   describe("An OrderItem", () => {
      let component: ShallowWrapper;

      const onDeleteOrderItem = jest.fn();

      const createProps = () => ({
         orderId: "o1",
         product: {
            id: "p1",
            description: "Product one",
            price: 3,
            category: ProductCategory.electronics,
         },
         productId: "p1",
         quantity: 3,
         onDeleteOrderItem,
      });

      beforeEach(() => {
         component = shallow(<OrderItemListItem {...createProps()}
         ></OrderItemListItem>);
      });

      it("should match snapshot", () => {
         expect(component.debug()).toMatchSnapshot();
      });

      it("should display the full product name", () => {
         expect(
            component.find(".orderitem-product-description").text(),
         ).toEqual("Product one");
      });

      it("should call onDeleteOrderItem", () => {
         component.find(".orderitem-button-delete").simulate("click");

         expect(onDeleteOrderItem).toBeCalled();
      });
   });

   describe("An OrderItem for an invalid product", () => {
      let component: renderer.ReactTestRenderer;

      beforeEach(() => {
         component = renderer.create(
            <Provider store={store}>
               <Router>
                  <ConnectedOrderItem productId="p2" orderId="o1" quantity={3} />
               </Router>
            </Provider>);
      });

      it("should match snapshot", () => {
         expect(component.toJSON()).toMatchSnapshot();
      });
   });
});
