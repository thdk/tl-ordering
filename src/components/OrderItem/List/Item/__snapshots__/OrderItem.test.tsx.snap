// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderItem rendering for a given state from redux store: An OrderItem for an invalid product should match snapshot 1`] = `null`;

exports[`OrderItem rendering for a given state from redux store: An OrderItem should match snapshot 1`] = `
"<li>
  <span className=\\"orderitem-product-description\\">
    Product one
  </span>
  :
  <span className=\\"orderitem-quantity\\">
    3
  </span>
   x $
  3.00
   = $
  9.00
  <input className=\\"orderitem-button-delete\\" type=\\"button\\" value=\\"x\\" onClick={[Function]} />
</li>"
`;
